dsl/src/main/scala/org/http4s/dsl/impl/Path.scala

```diff
@@ -218,6 +219,74 @@ object LongVar extends PathVar(str => Try(str.toLong))
   */
 object UUIDVar extends PathVar(str => Try(java.util.UUID.fromString(str)))
 
+/** Matrix path variable extractor
+  * For an example see [[https://www.w3.org/DesignIssues/MatrixURIs.html MatrixURIs]]
+  * This is useful for representing a resource that may be addressed in multiple dimensions where order is unimportant
+  *
+  * {{{
+  *
+  *    object BoardVar extends MatrixVar("square", List("x", "y"), str => Try(str.toInt).toOption)
+  *    Path("/board/square;x=5;y=3") match {
+  *      case Root / "board" / BoardVar(x, y) => ...
+  *    }
+  * }}}
+  */
+abstract class MatrixVar[F[_]: Traverse, A](
+    name: String,
+    domain: F[String],
+    range: String => Option[A]) {
```

Comment:
In common use, are all parameters in the matrix of the same type?

```
/address;city=Indianapolis;zipCode=46268
```

Do I have to represent `zipCode` as a `String`?  If we stripped `range` and returned `String`s, could we nest extractors?

```
case Root / AddressVar(city, Zip(zipCode)) =>
```

Reply:
I'm not really sure there is a common use. The problem I was specifically looking for was a means to solve awkward urls like /board/x/0/y/1 which implies a hierarchical structure of resources where none exists. Beyond that the goal is something that more people than just me is an ergonomic API. range is basically a mini-extractor so I will make this change.